### ISSUE TITLE
No autorestart on poor internet connectivity

### ISSUE THREAD

**Issue by Yury-MonZon** [whittlem/pycryptobot#199]
**Describe the bug**
After few retries the bot didn't reconnect.

**To Reproduce**
Run the bot, then drop your connection a few times.

**Expected behavior**
Bot should reconnect indefinitely.

**Screenshots**
`python3 pycryptobot.py --market BCHUSDT --verbose 1 --logfile ./logs/rbch.log --buypercent 33 --live 1 --disablebullonly --disablebuynearhigh --disablebuyobv --disablebuyelderray
--------------------------------------------------------------------------------
|                             Python Crypto Bot                                |
--------------------------------------------------------------------------------
|               Release : v2.1.0                                               |
--------------------------------------------------------------------------------
|                Market : BCHUSDT                                              |
|           Granularity : 1h                                                   |
--------------------------------------------------------------------------------
|              Bot Mode : LIVE - live trades using your funds!                 |
|           Bot Started : 2021-05-16 15:27:57.906531                           |
================================================================================
|          Sell At Loss : False  --sellatloss 0                                |
|    Sell At Resistance : True  --sellatresistance                             |
|       Trade Bull Only : False  --disablebullonly                             |
|         Buy Near High : False  --disablebuynearhigh                          |
|          Use Buy MACD : True  --disablebuymacd                               |
|           Use Buy OBV : False  --disablebuyobv                               |
|     Use Buy Elder-Ray : False  --disablebuyelderray                          |
|    Sell Fibonacci Low : True  --disablefailsafefibonaccilow                  |
|  Candlestick Reversal : True  --disableprofitbankreversal                    |
|              Telegram : True  --disabletelegram                              |
|                   Log : True  --disablelog                                   |
|               Tracker : True  --disabletracker                               |
|      Auto restart Bot : True  --autorestart                                  |
================================================================================
`
SKIPPED
`2021-05-16 23:03:04 | BCHUSDT | 1h | Current Price: 1145.6
2021-05-16 23:05:09 | BCHUSDT | 1h | Current Price: 1140.06
2021-05-16 23:07:14 | BCHUSDT | 1h | Current Price: 1146.97
2021-05-16 23:09:20 | BCHUSDT | 1h | Current Price: 1139.94
ReadTimeout(ReadTimeoutError("HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=20)"))
Traceback (most recent call last):
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 1062, in main
    runApp()
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 1059, in runApp
    s.run()
  File "/usr/local/lib/python3.9/sched.py", line 151, in run
    action(*argument, **kwargs)
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 325, in executeJob
    ticker = app.getTicker(app.getMarket())
  File "/home/debian/Documents/CryptoBot/pycryptobot/models/PyCryptoBot.py", line 353, in getTicker
    return api.getTicker(market)
  File "/home/debian/Documents/CryptoBot/pycryptobot/models/exchange/binance/api.py", line 443, in getTicker
    resp = self.client.get_symbol_ticker(symbol=market)
AttributeError: 'PublicAPI' object has no attribute 'client'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 382, in _make_request
    self._validate_conn(conn)
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 1010, in _validate_conn
    conn.connect()
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connection.py", line 411, in connect
    self.sock = ssl_wrap_socket(
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 428, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 472, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/local/lib/python3.9/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/local/lib/python3.9/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/usr/local/lib/python3.9/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
socket.timeout: _ssl.c:1112: The handshake operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/debian/.local/lib/python3.9/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/util/retry.py", line 532, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/packages/six.py", line 735, in reraise
    raise value
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 385, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=conn.timeout)
  File "/home/debian/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 336, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=20)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 1099, in <module>
    main()
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 1078, in main
    runApp()
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 1059, in runApp
    s.run()
  File "/usr/local/lib/python3.9/sched.py", line 151, in run
    action(*argument, **kwargs)
  File "/home/debian/Documents/CryptoBot/pycryptobot/pycryptobot.py", line 1013, in executeJob
    account.saveTrackerCSV(app.getMarket())
  File "/home/debian/Documents/CryptoBot/pycryptobot/models/TradingAccount.py", line 255, in saveTrackerCSV
    df = self.getOrders(market, '', 'done')
  File "/home/debian/Documents/CryptoBot/pycryptobot/models/TradingAccount.py", line 97, in getOrders
    model = BAuthAPI(self.app.getAPIKey(), self.app.getAPISecret(), self.app.getAPIURL())
  File "/home/debian/Documents/CryptoBot/pycryptobot/models/exchange/binance/api.py", line 68, in __init__
    self.client = Client(self.api_key, self.api_secret, { 'verify': False, 'timeout': 20 })
  File "/home/debian/.local/lib/python3.9/site-packages/binance/client.py", line 105, in __init__
    self.ping()
  File "/home/debian/.local/lib/python3.9/site-packages/binance/client.py", line 392, in ping
    return self._get('ping', version=self.PRIVATE_API_VERSION)
  File "/home/debian/.local/lib/python3.9/site-packages/binance/client.py", line 237, in _get
    return self._request_api('get', path, signed, version, **kwargs)
  File "/home/debian/.local/lib/python3.9/site-packages/binance/client.py", line 202, in _request_api
    return self._request(method, uri, signed, **kwargs)
  File "/home/debian/.local/lib/python3.9/site-packages/binance/client.py", line 196, in _request
    self.response = getattr(self.session, method)(uri, **kwargs)
  File "/home/debian/.local/lib/python3.9/site-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/home/debian/.local/lib/python3.9/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/debian/.local/lib/python3.9/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/home/debian/.local/lib/python3.9/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=20)`

**Desktop (please complete the following information):**
 - OS: Debian 10
 - Version 2.1.0

**Comment by whittlem** [comment]
I'm not really sure what I can do about this. It's recommended to have a stable connection for the bot as it needs to read pricing data every 2 minutes. I recommend deploying your bot(s) in the cloud. I'm going to be writing a Medium article on how to setup bots or free in Amazon AWS on EC2. Follow me on https://whittle.medium.com and it should be the next article you see from me.

**Comment by Yury-MonZon** [comment]
Well, reconnecting can be a solution =) I used to fetch data from Binance using binance.websockets and what I found that it also doesn't reconnect sometimes. So I switched to unicorn_binance_websocket_api.unicorn_binance_websocket_api_manager and it is rock solid now! Another way could be locking the bot inside of the infinite bash loop, which will restart it no matter what. But it is not elegant in any way.
